enhancements/agent-installer/automated-workflow-for-agent-based-installer.md

```diff
@@ -0,0 +1,624 @@
+the assisted-service API calls). The intermediate format will be
+considered another layer of control for the cluster creator, just as
+the manifests written by the `create manifests` sub-command are. The
+cluster creator will be able to generate those intermediate files with
+the sub-command `agent create-manifests`.
+
+The image generated by the `agent create-image` command will also
+contain one-time generated information, such as the InfraEnv ID needed
+by the agents to join a cluster deployment and REST API credentials
+needed for the agents and other clients to communicate securely with
+the assisted service.
+
+One host, designated as "node0", will be identified by the user
+through its IP address and configured to run the assisted service and
+a client that uses the assisted service REST API to automate the
+deployment. Because we must support single-node and compact
+deployments, the node selected as node0 must always be configured as a
+control plane node. Selecting other control plane nodes may be left to
+the user.
+
+All hosts will run the assisted agent, configured to talk to the
+service on node0.
+
+When all of the hosts are booted with the image, the client on node0
+will use the assisted service to trigger the deployment of the cluster
+(including applicable pre-flight validations) without further
+interaction from the user.
+
+The `agent create-image` command will produce as output credentials to
+access the cluster that is about to be created. This allows a client,
+such as a tool monitoring deployment progress, to use the cluster's
+API as soon as the API service is available.
+
+During deployment, the third-party orchestration tool will need a way
+to monitor progress and discover errors. The assisted service
+read-only API endpoints running on node0 will report that data. In the
+event of failure, a read-only endpoint in that service will be
+available to download any logs that have been collected.
+
+We will be taking advantage of the assisted service's ability to run
+the bootstrap services "in place" to avoid having a separate bootstrap
+VM. Bootstrap will run on node0, where the assisted service runs, to
+ensure that the service and its API are available for as long as
+possible.
+
+When bootstrapping completes and node0 reboots, progress monitoring
+will need to shift to the OpenShift APIs (`ClusterVersion` and
+`ClusterOperator`). We will extend the `openshift-install` binary with
+a new sub-command, `agent wait-for-install-complete`, to use both APIs for monitoring
+and [progress
+reporting](https://github.com/openshift/assisted-service/blob/master/docs/enhancements/installation-progress-bar.md)
+in that case, including reporting when steps like validation fail and
+the cluster deployment is never going to succeed. *Users will still be
+able to write their own monitoring logic using the assisted service
+and OpenShift APIs directly, but they will have the option of using
+the command line tool.*
```

Comment:
On a second thought, I think we could maintain the same command structure of the installer for the `wait-for`. Currently the `wait-for` has two sub-commands:

* `bootstrap-complete` Wait until cluster bootstrapping has completed
* `install-complete`   Wait until the cluster is ready

In the agent based workflow there's a bootstrap that could be potentially monitored (until node0 reboots) to satisfy the first sub-command. 
In this way we could keep more consistent the exposed interface for the end user, by offering a similar set of commands (and still continue to be open to changes)

Reply:
I think that makes sense. Is there already a pattern of commands with more than 3 components? We would have `agent wait-for install-complete`, for example.

Reply:
Good point, AFAICS the existing commands have a max depth level of two (base command / sub-command). Technically speaking the cobra framework shouldn't provide any limitation on this, so the `agent` base command could be used to provide a firm context for the subsequent commands, which at this point could mimic the usual look'n'feel. Ie, as a side-by-side comparison it could be something like that:

| openshift-install  | openshift-install agent |
| ------------- | ------------- |
| create manifests  | create manifests  |
| create cluster | - |
| - | create image |
| wait-for bootstrap-complete | wait-for bootstrap-complete |
| wait-for install-complete | wait-for install-complete |

Reply:
I've updated this document to use `wait-for install-complete` as we discussed in the meeting today. We can describe other commands in future enhancements when we decide which ones we need.